packages/flutter_tools/lib/src/cache.dart

```diff
@@ -1487,3 +1388,134 @@ const List<List<String>> _dartSdks = <List<String>> [
+    @required Net net,
+    @required Logger logger,
+    @required FileSystem fileSystem,
+    @required Directory tempStorage
+  }) : _operatingSystemUtils = operatingSystemUtils,
+       _net = net,
+       _logger = logger,
+       _fileSystem = fileSystem,
+       _tempStorage = tempStorage;
+
+  final Logger _logger;
+  final Net _net;
+  final OperatingSystemUtils _operatingSystemUtils;
+  final FileSystem _fileSystem;
+  final Directory _tempStorage;
+
+  /// Keep track of the files we've downloaded for this execution so we
+  /// can delete them after completion. We don't delete them right after
+  /// extraction in case [update] is interrupted, so we can restart without
+  /// starting from scratch.
+  @visibleForTesting
+  final List<File> downloadedFiles = <File>[];
+
+  /// Download a zip archive from the given [url] and unzip it to [location].
+  Future<void> downloadZipArchive(
+    String message,
+    Uri url,
+    Directory location,
+  ) {
+    return _downloadArchive(
+      message,
+      url,
+      location,
+      _operatingSystemUtils.unzip,
+    );
+  }
+
+  /// Download a gzipped tarball from the given [url] and unpack it to [location].
+  Future<void> downloadZippedTarball(String message, Uri url, Directory location) {
+    return _downloadArchive(
+      message,
+      url,
+      location,
+      _operatingSystemUtils.unpack,
+    );
+  }
+
+  /// Download an archive from the given [url] and unzip it to [location].
+  Future<void> _downloadArchive(
+    String message,
+    Uri url,
+    Directory location,
+    void Function(File, Directory) extractor,
+  ) async {
+    final String downloadPath = flattenNameSubdirs(url, _fileSystem);
+    final File tempFile = _createDownloadFile(downloadPath);
+    final Status status = _logger.startProgress(message, timeout: const Duration(minutes: 2));
+    int retries = 0;
+
+    while (retries < 2) {
+      try {
+        _ensureExists(tempFile.parent);
+        await _net.fetchUrl(url, destFile: tempFile, maxAttempts: 2);
+      } finally {
+        status.stop();
+      }
+      _ensureExists(location);
+      try {
+        extractor(tempFile, location);
+      } on ProcessException {
+        if (retries >= 1) {
+          rethrow;
+        }
+        retries += 1;
+        if (tempFile.existsSync()) {
+          tempFile.deleteSync();
+        }
+        continue;
+      }
+      return;
+    }
+  }
+
+  /// Create a temporary file and invoke [onTemporaryFile] with the file as
+  /// argument, then add the temporary file to the [downloadedFiles].
+  File _createDownloadFile(String name) {
+    final File tempFile = _fileSystem.file(_fileSystem.path.join(_tempStorage.path, name));
+    downloadedFiles.add(tempFile);
+    return tempFile;
+  }
+
+  /// Create the given [directory] and parents, as necessary.
+  void _ensureExists(Directory directory) {
+    if (!directory.existsSync()) {
+      directory.createSync(recursive: true);
```

Comment:
Will permissions errors on this operation be handled by the ErrorHandlingFileSystem?

Reply:
Updated the error handling filesystem to wrap Directory.createSync